src/packages/version-check/src/index.ts

```diff
@@ -0,0 +1,291 @@
+import { TruffleColors } from "@ganache/colors";
+import http2 from "http2";
+import Conf from "conf";
+import { default as semverDiff } from "semver/functions/diff";
+import { default as semverValid } from "semver/functions/valid";
+import { default as semverClean } from "semver/functions/clean";
+import { default as semverGte } from "semver/functions/gte";
+
+export type VersionCheckConfig = {
+  packageName: string;
+  enabled: boolean;
+  url: string;
+  ttl: number;
+  latestVersion: string;
+  latestVersionLogged: string;
+};
+
+export class VersionCheck {
+  protected ConfigManager;
+  protected _config;
+  protected _logger;
+  protected _currentVersion;
+
+  constructor(
+    currentVersion: string,
+    config?: VersionCheckConfig,
+    logger?: Function
+  ) {
+    // Creates a new config, or reads existing from disk
+    this.ConfigManager = new Conf({
+      configName: process.env.TEST ? "testConfig" : "config", // config is the package default
+      defaults: {
+        ...VersionCheck.DEFAULTS,
+        ...config
+      }
+    });
+
+    // pulls the config out of the manager, lays optional props over top
+    this._config = { ...this.ConfigManager.get("config"), ...config };
+
+    // If config was passed in, save changes
+    if (config) this.saveConfig();
+
+    if (this.isValidSemver(currentVersion)) {
+      this._currentVersion = this.cleanSemver(currentVersion);
+    } else {
+      // Semver is invalid, turn off version check
+      this._currentVersion = false;
+      this._config.enabled = false;
+    }
+
+    this._logger = logger || console;
+  }
+
+  init() {
+    this.getLatestVersion();
+    return this;
+  }
+
+  cleanSemver(semver) {
+    return semverClean(semver);
+  }
+
+  isValidSemver(semver) {
+    return semverValid(semver);
+  }
+
+  setEnabled(enabled) {
+    this.set("enabled", enabled);
+  }
+
+  setLatestVersion(latestVersion) {
+    if (this.isValidSemver(latestVersion)) {
+      this.set("latestVersion", latestVersion);
+    }
+  }
+
+  setLatestVersionLogged(latestVersionLogged) {
+    this.set("latestVersionLogged", latestVersionLogged);
+  }
+
+  setPackageName(packageName) {
+    this.set("packageName", packageName);
+  }
+
+  setTTL(ttl) {
+    this.set("ttl", ttl);
+  }
+
+  setUrl(url) {
+    this.set("url", url);
+  }
+
+  private set(key, value) {
+    this._config[key] = value;
+    this.saveConfig();
+  }
+
+  private saveConfig() {
+    this.ConfigManager.set("config", this._config);
+  }
+
+  configFileLocation() {
+    return this.ConfigManager.path;
+  }
+
+  // Intentionally verbose here if we get logging involved it could aid debugging
+  canNotifyUser() {
```

Comment:
Should we log if we are running in CI (can be easily detected with https://github.com/npm/ci-detect)?

Reply:
I don't think we should put any sort of funnel on it for now. 

`ci-detect` just sniffs a bunch of env vars. It wouldn't be hard to maintain our own list but do we want a piece of code that will sniff the env vars, at most, once a day? I am not inclined in favor of that.

Reply:
In CI it won't check them once per day, it'll check it on every run. I don't think these warnings are generally useful in a CI context (well... they definitely _could_ be read by someone that Ganache has an update available).

Why don't you think we should restrict this to local use only?